Example for FLASH nexus conversion/metadata collection

rettigl opened the pull request:
This PR is intended for discussing necessary/missing changes to the FLASH metadata structure for conversion to nexus.

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -8,9 +8,10 @@
+  "/ENTRY[entry]/entry_identifier": "@attrs:metadata/pid",
   "/ENTRY[entry]/USER[user]": {
     "name": "@attrs:metadata/principalInvestigator",
+    "affiliation": "t.b.d.",
```
The affiliation should be provided to uniquely identify a person. Can this be obtained from SciCat?

kutnyakhov replied:
This is still in the status to be filled manually, but as soon as DESY moves to the new version of SciCat this field should be added automatically from the beamtime metadata file.

rettigl replied:
You can add additional metadata by providing a dictionary, as done in example 3. Additionally, this can be read off from a yaml file. Maybe use one of these approaches as demonstration for missing metadata at the moment.

kutnyakhov replied:
In this particular case, if I am trying to add missing information as a manual metadata following tutorial 3
```
metadata = {}
metadata['principalInvestigator'] = {}
metadata['principalInvestigator']['affiliation'] = 'Deutsches Elektronen-Synchrotron DESY'
metadata['principalInvestigator']['email'] = 'dmytro.kutnyakhov@desy.de'
```
I am getting an error message:
ValueError: Cannot merge dictionary with non-dictionary entry!

For sure, I can add this info as e.g. a new key [user] or so, and then adjust config file accordingly using information from two different keys
```
"/ENTRY[entry]/USER[user]": {
    "name": "@attrs:metadata/principalInvestigator",
    "affiliation": "@attrs:metadata/user/affiliation",
```
but this way probably would a bit misleading.

rettigl replied:
Well, as the entry "principalInvestigator" is given as string by SciCat, you obviously cannot merge it with a dict, it can either be a dict or a string. 
If you don't like the option you sketched below (which I would find okay until these details are also in scicat), you can change the that gathers the data from Scicat, and rewrite the prinicpalInvestigator entry to be a dict, and the given string as "name" in that dict.

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -8,9 +8,10 @@
     "name": "@attrs:metadata/principalInvestigator",
+    "affiliation": "t.b.d.",
     "role": "Principal investigator",
     "email": "@attrs:metadata/contactEmail"
```
This contact email appears to be from the Data Owner, and not from the Principal Investigator. Is his/her email available?

kutnyakhov replied:
Same as before - in the current version of SciCat they are cutting a lot of info from beamtime metadata and not porting them into SciCat. This should be changed in the new version. At least the leader and PI will have their own affiliations and contact details.

rettigl replied:
Then add manually for now.

kutnyakhov replied:
Directly in the beamtime folder we have access to the beamtime_metadata.json file which is used to populate some of SciCat fields and in principle we can read that file and use that info, but then some fields will be duplicated several times.

rettigl replied:
How about you only take the fields that are not in SciCat from there?

rettigl replied:
I put those information into the notebook for now.

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -54,34 +55,32 @@
       "ky",
       "E"
     ],
-    "slow_axes": "@attrs:metadata/instrument/analyzer/slow_axes",
+    "slow_axes": "t.b.d.",
```
The only possible slow axis for FLASH is probably only "delayStage" at the moment, or empty for static scans. This should ideally be filled programatically by the loader or calibration routines.

kutnyakhov replied:
Probably also in the case of spin-resolved measurements switching magnetic coil could be another slow axes.

rettigl replied:
Maybe yes, spin-resolved ARPES is something still to develop completely.

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -54,34 +55,32 @@
-    "slow_axes": "@attrs:metadata/instrument/analyzer/slow_axes",
+    "slow_axes": "t.b.d.",
     "energy_resolution": {
-      "resolution": "@attrs:metadata/instrument/analyzer/energy_resolution",
+      "resolution": 100,
```
Can these be entered into SciCat, or obtained in some other means?

kutnyakhov replied:
Yes, this can go into SciCat and was already added into ingesting script during last beamtime

rettigl replied:
I see. Then maybe add a corresponding link here, or add manually.

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -54,34 +55,32 @@
   },
   "/ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]": {
     "projection": "@attrs:metadata/scientificMetadata/Collection/projection",
-    "scheme": "None",
+    "scheme": "momentum dispersive",
```
Should/could also be entered automatically depending on lens mode.

kutnyakhov replied:
This also added automatically into ScIcat during last beamtime

rettigl replied:
Then get it from sciCat

kutnyakhov replied:
In the description for NXcollectioncolumn/scheme, it has these options:
standard, deflector, PEEM, momentum microscope, etc.
For us only PEEM, momentum microscope valid or?
Can I am anyway use momentum dispersive as it was already ingested during last beamtimes or do I need to convert in one of those above?

rettigl replied:
This is outdated, in the new definitions it is only angular dispersive, momentum dispersive, non-dispersive or spatial dispersive:
https://fairmat-nfdi.github.io/nexus_definitions/classes/contributed_definitions/NXmpes.html#nxmpes
This should still be reflected in the base class, as I just noticed.

kutnyakhov replied:
Ok,
I was looking directly for NXcollectioncolumn
https://fairmat-nfdi.github.io/nexus_definitions/classes/contributed_definitions/NXcollectioncolumn.html#nxcollectioncolumn
and it still has old options.

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -54,34 +55,32 @@
     "lens_mode": "@attrs:metadata/scientificMetadata/Collection/lens_mode",
     "extractor_voltage": "@attrs:metadata/scientificMetadata/Collection/extractor_voltage",
     "extractor_voltage/@units": "V",
-    "extractor_current": "t.b.d.",
-    "extractor_current/@units": "µA",
```
If this is not available right now, just remove.

kutnyakhov replied:
For the moment this is tracked via Aux channels together with extractor voltage, thus we didn't send to SciCat but in principal this can be added into SciCat

rettigl replied:
You can get it then during metadata collection from the dataframe.

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -137,20 +130,18 @@
     "distance/@units": "mm",
     "incident_energy": "@attrs:metadata/scientificMetadata/Source/photon_energy/value",
-    "incident_energy/@units": "@attrs:metadata/scientificMetadata/Source/dispersion/unit",
-    "incident_energy_dispersion": "@attrs:metadata/scientificMetadata/Source/dispersion/value",
-    "incident_energy_dispersion/@units": "@attrs:metadata/scientificMetadata/Source/dispersion/unit",
```
I am not so sure what this actually means. Is this a property of the grating? There is a base class NXgrating for grating monochromators, maybe this can be used here.

kutnyakhov replied:
Yes, this is coming directly as beamline/grating metadata which we are reading in eV/mm - providing the bandwidth for the corresponding exit slit width.

rettigl replied:
This one I don't know right now, but I will discuss where this can go or if we need a new field.

rettigl replied:
Do you know what the propper name for this is, and e.g. if it is related to any concept in an ontology?

rettigl replied:
We have added a respective concept here: https://github.com/FAIRmat-NFDI/nexus_definitions/pull/257
Still need to find its way into the next release of pynxtools

rettigl replied:
done now.

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -137,20 +130,18 @@
-    "incident_energy_dispersion/@units": "@attrs:metadata/scientificMetadata/Source/dispersion/unit",
-    "pulse_duration": "t.b.d.",
+    "incident_energy/@units": "@attrs:metadata/scientificMetadata/Source/photon_energy/unit",
+    "pulse_duration": 100,
     "pulse_duration/@units": "fs",
```
Can be also removed here if unknown.

kutnyakhov replied:
This is a bit unfortunate - this value comes from LOLA measurements but quite often this device is broken and not delivering any data :(

rettigl replied:
Then enter manually in such a case?

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -137,20 +130,18 @@
     "pulse_duration/@units": "fs",
-    "incident_polarization": "horizontal",
-    "incident_polarization/@units": "None",
+    "incident_polarization": [1, 1, 0, 0],
+    "incident_polarization/@units": "V^2/mm^2",
```
Polarization as stokes vectors. Should be converted from SciCat values if possible.

kutnyakhov replied:
We have just the angle of the quarter wave plate, where e.g. 0 degrees = s-polarization, 45 degrees = p-polarization. How can I get to those vectors? :)

rettigl replied:
This is also something you can do in metadata generation from the WP angle.
This is maybe helpful.
For horizontal, it is (1, 1, 0, 0,), for vertical (1, -1, 0, 0), for any linear angle (1, cos2Theta, sin2Theta, 0), where Theta is the linear rotation angle of the light (twice the WP angle). 
For combined l/2 l/4 plates it becomes more complcated.

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -159,25 +150,27 @@
-    "pulse_duration": "@attrs:metadata/scientificMetadata/Laser/wavelength/pulse duration/value",
-    "pulse_duration/@units": "@attrs:metadata/scientificMetadata/Laser/wavelength/pulse duration/unit",
-    "incident_polarization": "@attrs:metadata/instrument/beam/pump/incident_polarization",
+    "incident_energy": "@attrs:metadata/scientificMetadata/Laser/energy/value",
+    "incident_energy/@units": "@attrs:metadata/scientificMetadata/Laser/energy/unit",
```
Calculate from wavelength, or add to SciCat

kutnyakhov replied:
So this is photon energy, 1.2 eV for 1030 nm, not a pulse energy?

rettigl replied:
see https://fairmat-nfdi.github.io/nexus_definitions/classes/base_classes/NXbeam.html#nxbeam
Yes, photon energy.
There is also pulse_energy

rettigl replied:
I put conversion into the notebook. If this layout is final, can also be added in the Flash metadata code.

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -159,25 +150,27 @@
+    "incident_energy/@units": "@attrs:metadata/scientificMetadata/Laser/energy/unit",
+    "pulse_duration": "@attrs:metadata/scientificMetadata/Laser/pulse duration/value",
+    "pulse_duration/@units": "@attrs:metadata/scientificMetadata/Laser/pulse duration/unit",
+    "incident_polarization": "@attrs:metadata/scientificMetadata/Laser/polarization",
```
Convert into stokes vector (what is the given number anyways?)

kutnyakhov replied:
This is number is angle of QWP

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -137,20 +130,18 @@
   },
-  "/ENTRY[entry]/INSTRUMENT[instrument]/source_TYPE[source_pump]": {
+  "/ENTRY[entry]/INSTRUMENT[instrument]/sourceTYPE[source_pump]": {
     "name": "OPCPA @ TR-ARPES @ FHI",
```
Update

rettigl commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -159,25 +150,27 @@
+  "/ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_pump]": {
     "distance": 0.0,
     "distance/@units": "mm",
     "incident_wavelength": "@attrs:metadata/scientificMetadata/Laser/wavelength/value",
```
This has the unit inside the "value" field in the scicat file, needs to be fixed there.

rettigl replied:
made a workaround

rettigl reviewed (commented):
Comments on the changes I did and TODOs

zain-sohail commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -0,0 +1,266 @@
+    "email": "@attrs:metadata/pi/email"
+  },
+  "/ENTRY[entry]/USER[user1]": {
+    "name": "@attrs:metadata/leader/last_name",
```
@kutnyakhov i changed from lastname to last_name here. Its because spelling check complains otherwise. You might need to adapt the tutorial

rettigl replied:
Anyways did not exist.

zain-sohail commented on sed/config/NXmpes_config-HEXTOF.json:
```diff
@@ -0,0 +1,266 @@
+  },
+  "/ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]": {
+    "depends_on": "@attrs:metadata/momentum_correction/registration/depends_on",
+    "TRANSFORMATIONS[transformations]": {
```
the spelling here of transformations was tranformations so fixed that too @kutnyakhov

zain-sohail reviewed (commented):
Dima sent me the updated nexus config file so I am just putting it here

rettigl commented:
@kutnyakhov  @zain-sohail Running the current state (rebased to master) outside maxwell results in the following warnings during saving:

```
Using mpes reader to convert the given files:  
• ../sed/config/NXmpes_config-HEXTOF.json
Could not find value for key /ENTRY[entry]/entry_identifier with value @attrs:metadata/_id.
Tried prefixes: [('@attrs', 'metadata/_id')].
Could not find value for key /ENTRY[entry]/USER[user0]/name with value @attrs:metadata/pi/last_name.
Tried prefixes: [('@attrs', 'metadata/pi/last_name')].
Could not find value for key /ENTRY[entry]/USER[user0]/affiliation with value @attrs:metadata/pi/institute.
Tried prefixes: [('@attrs', 'metadata/pi/institute')].
Could not find value for key /ENTRY[entry]/USER[user0]/email with value @attrs:metadata/pi/email.
Tried prefixes: [('@attrs', 'metadata/pi/email')].
Could not find value for key /ENTRY[entry]/USER[user1]/name with value @attrs:metadata/leader/last_name.
Tried prefixes: [('@attrs', 'metadata/leader/last_name')].
Could not find value for key /ENTRY[entry]/USER[user1]/affiliation with value @attrs:metadata/leader/institute.
Tried prefixes: [('@attrs', 'metadata/leader/institute')].
Could not find value for key /ENTRY[entry]/USER[user1]/email with value @attrs:metadata/leader/email.
Tried prefixes: [('@attrs', 'metadata/leader/email')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/projection with value @attrs:metadata/scientificMetadata/collectioncolumn/projection.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/projection')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/lens_mode with value @attrs:metadata/scientificMetadata/collectioncolumn/lens_mode.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/lens_mode')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/extractor_voltage with value @attrs:metadata/scientificMetadata/collectioncolumn/extractor_voltage/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/extractor_voltage/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/extractor_voltage/@units with value @attrs:metadata/scientificMetadata/collectioncolumn/extractor_voltage/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/extractor_voltage/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/extractor_current with value @attrs:metadata/scientificMetadata/collectioncolumn/extractor_current/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/extractor_current/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/extractor_current/@units with value @attrs:metadata/scientificMetadata/collectioncolumn/extractor_current/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/extractor_current/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/field_aperture/shape with value @attrs:metadata/scientificMetadata/collectioncolumn/field_aperture_shape.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/field_aperture_shape')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/field_aperture/size with value @attrs:metadata/scientificMetadata/collectioncolumn/field_aperture_size/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/field_aperture_size/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/field_aperture/size/@units with value @attrs:metadata/scientificMetadata/collectioncolumn/field_aperture_size/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/field_aperture_size/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/field_aperture/POSITIONER[fa_m1]/value with value @attrs:metadata/scientificMetadata/collectioncolumn/field_aperture_x/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/field_aperture_x/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/field_aperture/POSITIONER[fa_m1]/value/@units with value @attrs:metadata/scientificMetadata/collectioncolumn/field_aperture_x/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/field_aperture_x/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/field_aperture/POSITIONER[fa_m2]/value with value @attrs:metadata/scientificMetadata/collectioncolumn/field_aperture_y/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/field_aperture_y/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/field_aperture/POSITIONER[fa_m2]/value/@units with value @attrs:metadata/scientificMetadata/collectioncolumn/field_aperture_y/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/field_aperture_y/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/contrast_aperture/size with value @attrs:metadata/scientificMetadata/collectioncolumn/contrast_aperture_size/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/contrast_aperture_size/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/COLLECTIONCOLUMN[collectioncolumn]/contrast_aperture/size/@units with value @attrs:metadata/scientificMetadata/collectioncolumn/contrast_aperture_size/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/contrast_aperture_size/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/ENERGYDISPERSION[energydispersion]/pass_energy with value @attrs:metadata/scientificMetadata/collectioncolumn/tof_voltage/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/tof_voltage/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/ENERGYDISPERSION[energydispersion]/pass_energy/@units with value @attrs:metadata/scientificMetadata/collectioncolumn/tof_voltage/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/collectioncolumn/tof_voltage/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/DETECTOR[detector]/amplifier_type with value @attrs:metadata/scientificMetadata/detector/amplifier_type.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/detector/amplifier_type')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/DETECTOR[detector]/detector_type with value @attrs:metadata/scientificMetadata/detector/detector_type.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/detector/detector_type')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/DETECTOR[detector]/sensor_pixels with value @attrs:metadata/scientificMetadata/detector/sensor_pixels.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/detector/sensor_pixels')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/DETECTOR[detector]/amplifier_bias with value @attrs:metadata/scientificMetadata/detector/amplifier_bias/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/detector/amplifier_bias/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/DETECTOR[detector]/amplifier_bias/@units with value @attrs:metadata/scientificMetadata/detector/amplifier_bias/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/detector/amplifier_bias/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/DETECTOR[detector]/amplifier_voltage with value @attrs:metadata/scientificMetadata/detector/amplifier_voltage/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/detector/amplifier_voltage/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/DETECTOR[detector]/amplifier_voltage/@units with value @attrs:metadata/scientificMetadata/detector/amplifier_voltage/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/detector/amplifier_voltage/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/DETECTOR[detector]/detector_voltage with value @attrs:metadata/scientificMetadata/detector/detector_voltage/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/detector/detector_voltage/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/DETECTOR[detector]/detector_voltage/@units with value @attrs:metadata/scientificMetadata/detector/detector_voltage/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/detector/detector_voltage/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/sourceTYPE[source_probe]/name with value @attrs:metadata/scientificMetadata/source/name.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/source/name')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/sourceTYPE[source_probe]/probe with value @attrs:metadata/scientificMetadata/source/probe.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/source/probe')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/sourceTYPE[source_probe]/type with value @attrs:metadata/scientificMetadata/source/type.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/source/type')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/sourceTYPE[source_probe]/frequency with value @attrs:metadata/scientificMetadata/source/repetition_rate/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/source/repetition_rate/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/sourceTYPE[source_probe]/frequency/@units with value @attrs:metadata/scientificMetadata/source/repetition_rate/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/source/repetition_rate/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_probe]/incident_energy with value @attrs:metadata/scientificMetadata/source/photon_energy/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/source/photon_energy/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_probe]/incident_energy/@units with value @attrs:metadata/scientificMetadata/source/dispersion/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/source/dispersion/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_probe]/incident_energy_dispersion with value @attrs:metadata/scientificMetadata/source/dispersion/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/source/dispersion/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_probe]/incident_energy_dispersion/@units with value @attrs:metadata/scientificMetadata/source/dispersion/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/source/dispersion/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_pump]/incident_wavelength with value @attrs:metadata/scientificMetadata/laser/wavelength/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/laser/wavelength/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_pump]/incident_wavelength/@units with value @attrs:metadata/scientificMetadata/laser/wavelength/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/laser/wavelength/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_pump]/incident_energy with value @attrs:metadata/scientificMetadata/laser/incident_energy/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/laser/incident_energy/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_pump]/incident_energy/@units with value @attrs:metadata/scientificMetadata/laser/incident_energy/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/laser/incident_energy/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_pump]/pulse_duration with value @attrs:metadata/scientificMetadata/laser/pulse_duration/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/laser/pulse_duration/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_pump]/pulse_duration/@units with value @attrs:metadata/scientificMetadata/laser/pulse_duration/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/laser/pulse_duration/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_pump]/incident_polarization with value @attrs:metadata/scientificMetadata/laser/incident_polarization/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/laser/incident_polarization/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_pump]/incident_polarization/@units with value @attrs:metadata/scientificMetadata/laser/incident_polarization/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/laser/incident_polarization/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_pump]/pulse_energy with value @attrs:metadata/scientificMetadata/laser/pulse_energy/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/laser/pulse_energy/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_pump]/pulse_energy/@units with value @attrs:metadata/scientificMetadata/laser/pulse_energy/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/laser/pulse_energy/unit')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/MANIPULATOR[manipulator]/type with value @attrs:metadata/scientificMetadata/manipulator/type.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/manipulator/type')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/MANIPULATOR[manipulator]/sample_bias_voltmeter/value with value @attrs:metadata/scientificMetadata/manipulator/sample_bias/value.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/manipulator/sample_bias/value')].
Could not find value for key /ENTRY[entry]/INSTRUMENT[instrument]/MANIPULATOR[manipulator]/sample_bias_voltmeter/value/@units with value @attrs:metadata/scientificMetadata/manipulator/sample_bias/unit.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/manipulator/sample_bias/unit')].
Could not find value for key /ENTRY[entry]/SAMPLE[sample]/name with value @attrs:metadata/scientificMetadata/sample/sample_name.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/sample/sample_name')].
Could not find value for key /ENTRY[entry]/SAMPLE[sample]/preparation_description with value @attrs:metadata/scientificMetadata/sample/preparation_description.
Tried prefixes: [('@attrs', 'metadata/scientificMetadata/sample/preparation_description')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/DISTORTION[distortion]/symmetry with value @attrs:metadata/momentum_correction/correction/rotsym.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/correction/rotsym')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/DISTORTION[distortion]/original_centre with value @attrs:metadata/momentum_correction/correction/pcent.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/correction/pcent')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/DISTORTION[distortion]/original_points with value @attrs:metadata/momentum_correction/correction/pouter.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/correction/pouter')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/DISTORTION[distortion]/cdeform_field with value @attrs:metadata/momentum_correction/correction/cdeform_field.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/correction/cdeform_field')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/DISTORTION[distortion]/rdeform_field with value @attrs:metadata/momentum_correction/correction/rdeform_field.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/correction/rdeform_field')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/depends_on with value @attrs:metadata/momentum_correction/registration/depends_on.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/depends_on')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[trans_x] with value @attrs:metadata/momentum_correction/registration/trans_x/value.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/trans_x/value')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[trans_x]/@transformation_type with value @attrs:metadata/momentum_correction/registration/trans_x/type.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/trans_x/type')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[trans_x]/@units with value @attrs:metadata/momentum_correction/registration/trans_x/units.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/trans_x/units')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[trans_x]/@vector with value @attrs:metadata/momentum_correction/registration/trans_x/vector.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/trans_x/vector')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[trans_x]/@depends_on with value @attrs:metadata/momentum_correction/registration/trans_x/depends_on.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/trans_x/depends_on')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[trans_y] with value @attrs:metadata/momentum_correction/registration/trans_y/value.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/trans_y/value')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[trans_y]/@units with value @attrs:metadata/momentum_correction/registration/trans_y/units.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/trans_y/units')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[trans_y]/@transformation_type with value @attrs:metadata/momentum_correction/registration/trans_y/type.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/trans_y/type')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[trans_y]/@vector with value @attrs:metadata/momentum_correction/registration/trans_y/vector.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/trans_y/vector')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[trans_y]/@depends_on with value @attrs:metadata/momentum_correction/registration/trans_y/depends_on.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/trans_y/depends_on')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[rot_z] with value @attrs:metadata/momentum_correction/registration/rot_z/value.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/rot_z/value')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[rot_z]/@units with value @attrs:metadata/momentum_correction/registration/rot_z/units.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/rot_z/units')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[rot_z]/@transformation_type with value @attrs:metadata/momentum_correction/registration/rot_z/type.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/rot_z/type')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[rot_z]/@offset with value @attrs:metadata/momentum_correction/registration/rot_z/offset.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/rot_z/offset')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[rot_z]/@vector with value @attrs:metadata/momentum_correction/registration/rot_z/vector.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/rot_z/vector')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/REGISTRATION[registration]/TRANSFORMATIONS[transformations]/AXISNAME[rot_z]/@depends_on with value @attrs:metadata/momentum_correction/registration/rot_z/depends_on.
Tried prefixes: [('@attrs', 'metadata/momentum_correction/registration/rot_z/depends_on')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/CALIBRATION[kx_calibration]/scaling with value @attrs:metadata/momentum_calibration/calibration/kx_scale.
Tried prefixes: [('@attrs', 'metadata/momentum_calibration/calibration/kx_scale')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/CALIBRATION[kx_calibration]/offset with value @attrs:metadata/momentum_calibration/calibration/x_center.
Tried prefixes: [('@attrs', 'metadata/momentum_calibration/calibration/x_center')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/CALIBRATION[kx_calibration]/calibrated_axis with value @attrs:metadata/momentum_calibration/calibration/kx_axis.
Tried prefixes: [('@attrs', 'metadata/momentum_calibration/calibration/kx_axis')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/CALIBRATION[ky_calibration]/scaling with value @attrs:metadata/momentum_calibration/calibration/ky_scale.
Tried prefixes: [('@attrs', 'metadata/momentum_calibration/calibration/ky_scale')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/CALIBRATION[ky_calibration]/offset with value @attrs:metadata/momentum_calibration/calibration/y_center.
Tried prefixes: [('@attrs', 'metadata/momentum_calibration/calibration/y_center')].
Could not find value for key /ENTRY[entry]/PROCESS[process]/CALIBRATION[ky_calibration]/calibrated_axis with value @attrs:metadata/momentum_calibration/calibration/ky_axis.
Tried prefixes: [('@attrs', 'metadata/momentum_calibration/calibration/ky_axis')].
The entry '/ENTRY[entry]/definition/@version' (value: v2022.07.post1.dev1490+gd3c5237d) should not be changed by the reader. The provided version 'None' is kept. We assume you know what you are doing.
WARNING: The data entry corresponding to /ENTRY[entry]/USER[user0]/name is required and hasn't been supplied by the reader.
WARNING: The data entry corresponding to /ENTRY[entry]/USER[user0]/affiliation is required and hasn't been supplied by the reader.
WARNING: The data entry corresponding to /ENTRY[entry]/USER[user1]/name is required and hasn't been supplied by the reader.
WARNING: The data entry corresponding to /ENTRY[entry]/USER[user1]/affiliation is required and hasn't been supplied by the reader.
WARNING: The data entry corresponding to /ENTRY[entry]/INSTRUMENT[instrument]/sourceTYPE[source_probe]/type is required and hasn't been supplied by the reader.
WARNING: The data entry corresponding to /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_probe]/incident_energy is required and hasn't been supplied by the reader.
WARNING: The data entry corresponding to /ENTRY[entry]/INSTRUMENT[instrument]/beamTYPE[beam_pump]/incident_energy is required and hasn't been supplied by the reader.
WARNING: The required group, /ENTRY[entry]/INSTRUMENT[instrument]/ELECTRONANALYSER[electronanalyser]/DETECTOR, hasn't been supplied.
WARNING: The data entry corresponding to /ENTRY[entry]/INSTRUMENT[instrument]/MANIPULATOR[manipulator]/sample_bias_voltmeter/value is required and hasn't been supplied by the reader.
WARNING: The data entry corresponding to /ENTRY[entry]/SAMPLE[sample]/name is required and hasn't been supplied by the reader.
WARNING: Field /ENTRY[entry]/SAMPLE[sample]/temperature/temperature_sensor written without documentation.
WARNING: Field /ENTRY[entry]/SAMPLE[sample]/gas_pressure/pressure_gauge written without documentation.
WARNING: Field /ENTRY[entry]/PROCESS[process]/energy_calibration/coefficients written without documentation.
WARNING: Field /ENTRY[entry]/PROCESS[process]/energy_calibration/fit_function written without documentation.
WARNING: Field /ENTRY[entry]/PROCESS[process]/energy_calibration/original_axis written without documentation.
WARNING: Field /ENTRY[entry]/PROCESS[process]/energy_calibration/calibrated_axis written without documentation.
The output file generated: runs44824-27.nxs.
```
I.e. a huge number of metadata is not being found.

Running the same on Maxwell results in an error when trying to load the metadata file:

```
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
Cell In[46], line 3
      1 from sed.core.config import load_config
      2 import numpy as np
----> 3 metadata = load_config(path + "/44824_20230324T060430.json")
      5 # Fix metadata
      6 metadata["scientificMetadata"]["Laser"]["wavelength"]["value"] = float(metadata["scientificMetadata"]["Laser"]["wavelength"]["value"][:-2])

File ~/sed/sed/core/config.py:159, in load_config(config_path)
    157 config_file = Path(config_path)
    158 if not config_file.is_file():
--> 159     raise FileNotFoundError(
    160         f"could not find the configuration file: {config_file}",
    161     )
    163 if config_file.suffix == ".json":
    164     with open(config_file, encoding="utf-8") as stream:

FileNotFoundError: could not find the configuration file: /asap3/flash/gpfs/pg2/2023/data/11019101/raw/hdf/offline/fl1user3/44824_20230324T060430.json
```

kutnyakhov commented:
@rettigl for FLASH data we are using the SciCat catalogue for the metadata. In order to get them you need to get your personal `SciCat Token` from `Settings` and add/update it into your config file under `metadata:` section as `scicat_token:`
or add that info manually:
```
config_override['metadata'] = {}
config_override['metadata'] = {
    'scicat_url': 'https://scicat-flash-test1.desy.de/api/v3',
    'scicat_token': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2Mzc0YmYyMmE4NWQ0N2QwMGE5MWRlN2MiLCJ1c2VybmFtZSI6Imt1dG55YWtkIiwiZW1haWwiOiJkbXl0cm8ua3V0bnlha2hvdkBkZXN5LmRlIiwiYXV0aFN0cmF0ZWd5IjoibG9jYWwiLCJpZCI6IjYzNzRiZjIyYTg1ZDQ3ZDAwYTkxZGU3YyIsInVzZXJJZCI6IjYzNzRiZjIyYTg1ZDQ3ZDAwYTkxZGU3YyIsImlhdCI6MTcyODU1MDc5NCwiZXhwIjoxNzI4NTkzOTk0fQ.Hqy6ICvBBMLMq8KwCLf-5tRsA-BdWGfMCR9mjWv4ldU'
}
```
Also, as discussed during last meeting, we still have problems that some user info is missing and IT guys are working to solve this problem for future, but for now we have to also read additionally beamtime metadata file:
```
f = open('/asap3/flash/gpfs/pg2//2023/data/11019101/beamtime-metadata-11019101.json', "r")
beamtime_metadata = json.loads(f.read())

beamtime_metadata_contact = {'leader': beamtime_metadata['leader'], 'pi': beamtime_metadata['pi']}
```
and accordingly, add this info while reading data
```
sp = SedProcessor(
    runs=[44824,44825,44826,44827],
    config=config_override,
    system_config=config_file,
    metadata=beamtime_metadata_contact,
    collect_metadata=True
)
```

rettigl commented:
> for FLASH data we are using the SciCat catalogue for the metadata. In order to get them you need to get your personal `SciCat Token` from `Settings` and add/update it into your config file under `metadata:` section as `scicat_token:`
> or add that info manually:

For general usage that is fine, but for the example we need to provide a solution that works without this, i.e. put the scicat metadata into a file and provide it with the example data. This exists in the zenodo upload, but not in the data on the maxwell cluster, thus it fails there.

> Also, as discussed during last meeting, we still have problems that some user info is missing and IT guys are working to solve this problem for future, but for now we have to also read additionally beamtime metadata file:

This file again is not part of the upload on zenodo, so again it would fail if you are not on maxwell

Even with these additions, there are a lot of metadata not being found by the reader, because the information in the config file is wrong. This should be fixed for this example, that is my point. If some data are not available for this, let's just remove the respective entries and add them later once these data are available.

rettigl commented:
@zain-sohail @kutnyakhov I removed everything that does not exist here, and corrected to reflect the paths in the metadata files.
The example still fails on maxwell because the metadata json files are not present there. I suggest to copy them there (I cannot do it), to make this work for now. If more metadata are readily available, this can still be added later.

zain-sohail commented on tutorial/4_hextof_workflow.ipynb:
```diff
@@ -863,6 +863,26 @@
+   "source": [
+    "from sed.core.config import load_config\n",
+    "import numpy as np\n",
+    "metadata = load_config(path + \"/44824_20230324T060430.json\")\n",
```
I have put the metadata file in the shared folder so this should work but is rather a crude way
`metadata = load_config(beamtime_dir + "/shared/44824_20230324T060430.json")`

rettigl replied:
Included
